source/extensions/priv/server/ntds_decrypt.c

```diff
@@ -0,0 +1,194 @@
+	}
+	cryptOK = decrypt_hash_from_rid(encHashData, &rid, decHash);
+	if (!cryptOK){
+		return FALSE;
+	}
+	bytes_to_string(decHash, HASH_LENGTH_BYTES, hashString);
+	return TRUE;
+}
+
+/*!
+* @brief Wraps SystemFunction025 which decrypts a hash using the RID
+* @param encodedHash Pointer to a byte array containing the encrypted hash
+* @param rid Pointer to a DWORD containing the Relative ID(RID)
+* @param decodedHash Pointer to where we store the decrypted hash
+* @returns Indication of sucess or failure.
+*/
+BOOL decrypt_hash_from_rid(LPBYTE encodedHash, LPDWORD rid, LPBYTE decodedHash){
+	typedef NTSTATUS(__stdcall *PSYS25)(IN LPCBYTE data, IN LPDWORD key, OUT LPBYTE output);
+	HMODULE hAdvapi = LoadLibrary("advapi32.dll");
+	if (hAdvapi == NULL){
+		return FALSE;
+	}
+	PSYS25 decryptFromRID = (PSYS25)GetProcAddress(hAdvapi, "SystemFunction025");
+	if (decryptFromRID(encodedHash, rid, decodedHash) != 0){
+		return FALSE;
+	}
+	return TRUE;
+}
+
+/*!
+* @brief Splits up an encrypted LM or NT hash history and decrypts each stored hash
+* @param encHashHistory Pointer to the byte array containing the hash history record
+* @param sizeHistory The size of the history record
+* @param pekDecrypted Pointer to a decryptedPEK structure that holds our decrypted PEK
+* @param rid DWORD representing the Relative ID(RID) of the account
+* @param accountHistory Pointer to a string wherewe store all the decrypted historical hashes
+* @param historyCount Pointer to n integer where we store a count of the historical hashes
+* @returns Indication of sucess or failure.
+*/
+BOOL decrypt_hash_history(LPBYTE encHashHistory, size_t sizeHistory, struct decryptedPEK *pekDecrypted, DWORD rid, char *accountHistory, int *historyCount){
+	BOOL cryptOK = FALSE;
+	int sizeHistoryData = (int)sizeHistory - 24;
+	int numHashes = (sizeHistoryData / HASH_LENGTH_BYTES);
+	memcpy(historyCount, &numHashes, sizeof(historyCount));
+	LPBYTE encHistoryData = (LPBYTE)calloc(1,sizeHistoryData);
+	LPBYTE decHistoryData = (LPBYTE)calloc(1,(sizeHistoryData * 2));
+	memcpy(encHistoryData, encHashHistory + 24, sizeHistoryData);
+	cryptOK = decrypt_rc4(pekDecrypted->pekKey, encHashHistory + 8, encHistoryData, 1, sizeHistoryData);
+	if (!cryptOK){
+		free(encHistoryData);
+		free(decHistoryData);
+		return FALSE;
+	}
+	LPBYTE historicalHash = encHistoryData;
+	LPBYTE writeMarker = decHistoryData;
+	for (int i = 0; i < numHashes; i++){
+		BYTE decHash[HASH_LENGTH_BYTES];
+		char hashString[NULL_TERIMNATED_HASH_STRING_LENGTH];
+		cryptOK = decrypt_hash_from_rid(historicalHash, &rid, decHash);
+		if (!cryptOK){
+			return FALSE;
+		}
+		bytes_to_string(decHash, HASH_LENGTH_BYTES, hashString);
+		strncpy(writeMarker, hashString, NULL_TERIMNATED_HASH_STRING_LENGTH);
+		historicalHash = historicalHash + HASH_LENGTH_BYTES;
+		writeMarker = writeMarker + NULL_TERIMNATED_HASH_STRING_LENGTH;
+	}
+	memcpy(accountHistory, decHistoryData, (numHashes * NULL_TERIMNATED_HASH_STRING_LENGTH));

```

Comment:
Again this mul in the length of a memcpy makes me a little bit afraid, if you've lengths under control you can forget about my worries, but just warning :)